src/io/merfishdata.rs

```diff
@@ -3,55 +3,260 @@
+        #[serde(rename = "distPeriphery")]
+        pub dist_periphery: f64,
+    }
 
-impl<R: io::Read> Reader<R> {
-    pub fn from_reader(rdr: R) -> Self {
-        Reader {
-            inner: csv::Reader::from_reader(rdr).delimiter(b'\t')
+    impl MerfishRecord for Record {
+        fn cell_id(&self) -> u32 {
+            self.cell_id
         }
+
+        fn cell_name(&self) -> String { self.cell_id.to_string() }
+
+        fn cell_pos(&self) -> (f32, f32) { (self.abs_position[0], self.abs_position[1]) }
+
+        fn feature_id(&self) -> u16 { self.barcode_id }
+
+        fn feature_name(&self) -> String { self.barcode_id.to_string() }
+
+        fn hamming_dist(&self) -> u8 { 1 - self.is_exact }
+    }
+
+    pub struct Reader<R: io::Read> {
+        reader: io::BufReader<R>,
+        header: Header,
+        item: usize,
     }
 
-    pub fn records<'a>(&'a mut self) -> csv::DecodedRecords<'a, R, Record> {
-        self.inner.decode()
+    impl<R: io::Read> Reader<R> {
+        /// Constructs a new `merfish::Reader<R>` where `R: io::Read`.
+        ///
+        /// This will read the header in order to check if the version matches.
+        pub fn new(reader: R) -> Result<Self, &'static str> {
+            let mut reader = io::BufReader::new(reader);
+            let header: Result<Header, _> = bincode::deserialize_from(&mut reader);
+            match header {
+                Ok(ref header) if header.version != 1 => Err("Unsupported version {}"),
+                Ok(ref header) if header.is_corrupt != 0 => Err("Header is corrupt, i.e. might not have been written properly (while in append mode)"),
+                Ok(header) => {
+                    // Read (discard) table header
+                    let mut header_buf = vec![0u8; header.header_length as usize];
+                    let r = reader.read_exact(&mut header_buf);
+                    match r {
+                        Err(_) => Err("Failed reading table header"),
+                        Ok(()) => Ok(Reader {
+                            reader,
+                            header,
+                            item: 0,
+                        })
+                    }
+                }
+                Err(_) => Err("Failed reading header")
+            }
+        }
+
+        pub fn header(&self) -> &Header {
+            &self.header
+        }
+    }
+
+    impl Reader<fs::File> {
+        /// Constructs a new `merfish::Reader<fs::File>` from given file.
+        ///
+        /// This delegates work to `merfish::Reader::new`.
+        pub fn from_file<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+            fs::File::open(path).map(|r| { Reader::new(r).unwrap() })
+        }
+    }
+
+    impl<R: io::Read> Iterator for Reader<R> {
+        type Item = io::Result<Record>;
+
+        /// Reads a single `merfish::Record` from the file, advancing the iterator.
+        fn next(&mut self) -> Option<<Self as Iterator>::Item> {
```

Comment:
What about this:

```
pub fn records(&self) -> impl Iterator<Item=io::Result<Record>> {
            (0..self.header.num_entries).map(|i| {
                match bincode::deserialize_from::<_, Record>(&mut self.reader) {
                    Ok(record) => {
                        Some(Ok(record))
                    }
                    // TODO: propagate bincode's Error
                    Err(_) => Some(
                        Err(io::Error::new(
                            io::ErrorKind::Other,
                            format!("Failed deserializing record {} from reader.", i)
                        ))
                    )
                }
            })
        }
```

(untested)

Reply:
Oh, I guess I could have thought of that :D 
It'd have to be `records(&mut self)`, also I just realized wrapping the Record in an `Option` isn't actually needed.
I guess this change will, however, make lifetime annotations necessary.

Reply:
This might work (not sure about moving `i`, but otherwise it'd outlive `self` [from self.header.num_entries]).

```
pub fn records<'a>(&'a mut self) -> impl Iterator<Item=io::Result<Record>> + 'a {
            (0..self.header.num_entries).map(move |i| {
                match bincode::deserialize_from::<_, Record>(&mut self.reader) {
                    Ok(record) => {
                        Ok(record)
                    }
                    // TODO: propagate bincode's Error
                    Err(_) =>
                        Err(io::Error::new(
                            io::ErrorKind::Other,
                            format!("Failed deserializing record {} from reader.", i), )
                        )
                }
            })
        }
```

Reply:
Are you sure about the `i`? It is a primitive with copy semantics...

Reply:
Was confused about move semantics for a second there; the problem is that `self` is borrowed via `&mut self.reader` in the closure body,
i.e. `closure may outlive the current function, but it borrows `self`, which is owned by the current function`.

Reply:
Ok, just commit the current version then we can play around with it before merging.